### ISSUE TITLE
Support SSH certs / ca functionality

### ISSUE THREAD

**Issue by Firstyear** [kanidm/kanidm#198]
Support kanidm acting as an SSH ca signer, which then allows clients to have policy based and ephemeral certs issued. The issuing process and this needs a lot of scoping out of course :)

**Comment by Firstyear** [comment]
note: policy/restricted command certs with a long/permanent life for service accounts ie. 

It should be possible to store keys with options like :

command="borg serve --restrict-to-repository /home/backup/",restrict

This way wecan store keys for service accounts centrally and potentially even hav policies on the keys command?

**Comment by csarn** [comment]
As this seems like a lot of effort, maybe it would be possible to first integrate with https://github.com/smallstep/certificates? This already provides a SSH CA, and can use OIDC. The SSH certificate can be templated, so a "restrict_command" OIDC claim could be integrated into the SSH cert, if kanidm provides that claim.

**Comment by Firstyear** [comment]
Hi @csarn 

In the Kanidm project we have some strict rules about external dependencies. These are covered in https://kanidm.github.io/kanidm/stable/developers/faq.html#why-dont-you-use-libraryproject-x 

Using smallstep would violate these both on being not-rust, but also including a microservice architecture. As a result it's a non-starter.

This is because of practical experience having seen what happens to real deployments of IDM's with microservices, the outages they cause and the difficulty in fault resolution. As well, as developers we must be able to test our project, and external dependencies make that process exponentially more complex.

So as a result, this feature will happens when we have time to focus on doing it *correctly* and *safely*. Nothing less will satisfy us.

Thanks,

**Comment by csarn** [comment]
With "integrate smallstep" I was thinking of integration on the operational/admin side. With existing OIDC features this should already be possible. I wasn't at all suggesting integrating actual smallstep-related code into kanidm :)
I read this issue as if it was a feature request by a kanidm user that just wants to also have SSH CA integration. For such a usecase, one could try smallstep until kanidm provides similar functionality.
Now I see that this is actually a roadmap item planned by a developer, so my suggestion is useless in that context.

**Comment by Firstyear** [comment]
Ahhh my apologies. I didn't understand your comment correctly then. If you just want to add this as an external integration in the doc, that would be okay. 

Regardless, sounds like there are some issues with smallstep right now though, so lets get that resolved first ....

**Comment by Gabgobie** [comment]
~Edit: After getting over the major disappointment that are "vanilla" SSH certificates and letting my mind linger for a while, I had another idea that I'll elaborate on as I find the time and that can actually improve on security compared to the current system.~

Edit 2: For that to work, `unixd` would need access to the supplied user certificate, which, to my understanding, we don't have.

---

TL;DR: If you want to pursue SSH certs _at all_, I suggest only doing host certs for now. This is probably blocked by domain joins. The long read contains reasoning and is the result of me trying to turn my disappointment into something productive.

L;R:

> The issuing process and this needs a lot of scoping out of course :)

I just tested out SSH certs with `ssh-keygen` and `sshd` and was severely disappointed to the point of reconsidering if SSH certs are something worth pursuing at all. So... I thought I'd look around the internet if I was missing something and document my findings here.

---

Going by the [IETF draft from 2025-10-25](https://www.ietf.org/archive/id/draft-miller-ssh-cert-01.html) DuckDuckGo surfaced, there are very few fields we can use to constrain/empower a certificate.

We get a handful of flag-type extensions that enable us to forbid specific ssh features. Extensions are never critical and MUST be ignored without error if they are not recognized making their meaningfulness questionable.

Other than that we have three critical options:

- `force-command` Forces execution of a command command upon login
- `source-address` probably the most useful one to make sure a certificate can only be used from one client but also only sensible under certain conditions (`short-lived + dhcp` / `static ip`).
- `verify-required` only applicable to FIDO-backed keys

Smallstep makes contradictory claims about SSH Cert RBAC being directly encoded and cryptographically enforced in the cert: <https://smallstep.com/product/ssh/#radix-_r_c_> (The link sucks a little because they don't have an ID on the overarching div, only the button)

> SSH certificates can carry rich authorization context directly inside the credential itself. They can encode allowed hosts, user principals, group-based privileges, sudo constraints, time-of-day restrictions, maximum session duration, and optionally device identity.

~I'll have to look into their templates/code/output to figure out what they are doing to encode/enforce these requirements. Maybe also take another look at `sshd` for recognized extensions.~

After glancing over <https://smallstep.com/docs/ssh/how-it-works/>, I find the claim

> They can encode allowed hosts, user principals, group-based privileges, sudo constraints, time-of-day restrictions, maximum session duration, and optionally device identity.

somewhat questionable. They seem to be encoding a principal and manage ACLs with a NSS service. In our case this would be handled by `unixd`.

---

SSH CAs suck even worse. From what I can tell, you just generate an SSH key pair, say ed25519, and declare this the CA. No constraints, no nothing. Then you just copy the public key to the server and tell it to accept any cert signed by the corresponding private key. <https://docs.rockylinux.org/10/guides/security/ssh_ca_key_signing/>

That makes the blast radius of a compromised SSH CA worse than a normal SSH key by a long shot as it can sign for any duration, any source and any principal - including root. A CA can be compromised silently! If a normal SSH key is compromised, the public counterpart must be somewhere in your infra and available to `sshd`. You can purge it and the compromised private key is useless. When using a CA, you MUST keep track of all issued keys to be able to revoke them. If your CA was accessed without authorization and a non-bounded key was issued without your knowledge, you have NO WAY to issue a KRL.

---

In contrast, caching aside, `kanidm-unixd` could implement fine grained ssh access operationally. Personally, I have a strong distaste for operational security - which requires a level of trust - and prefer cryptographic restrictions. (That's why I have invested/am investing a significant amount of time into learning x.509 to use a severely constrained CA that cannot issue valid certs outside its designated domain.) Given the conditions outlined above, we are dealing with operational security anyway and plain ssh keys with kani-supplied capabilities and validity constraints seems to be the better way right now.

---

Finally, I am a bit torn on this. SSH certs are the only real way for me to issue a key with limited validity. At the same time they seem to introduce risks that are beyond anything acceptable. Maybe I'm even being unjust due to my disappointment and frustration - you be the judge. My excitement certainly created some high expectations for SSH certs and these expectations were not met.

---

Keeping this productive, I'd like to end on something actionable.

Plain/Classical(?) SSH user certs would by my account and to my surprise lower security. BUT that doesn't mean all is lost. `TrustedUserCAKeys` points to a file that has one ssh public key per line. I just tested and confirmed that multiple keys may be in there and user certificates signed by any one of them are accepted. That means we _could_ register a separate short lived CA per user so users only ever get a key with a bounded lifetime. I still struggle to see the value here, especially given the offline requirements and caching/invalidation difficulties.

Additionally, everything above mostly concerns _user_ certs. The SSH TOFU issue can still be solved by certs! While the SSH CA does not have a certificate and thus no concept like x.509 name constraints, the user has to add it (the public key) to the known hosts file and defines the [scope of validity](https://docs.rockylinux.org/10/guides/security/ssh_ca_key_signing/#creating-a-certificate-authority) there.

---

Did I say finally earlier? Promise, this one is the last section - and it's a short one! My recommendation: SSH host certificates seem to be something worth pursuing or at least nice to have but likely require the planned domain join feature. User certs not so much.

Our TODOs would be

- Implement SSH signing capabilities in kani
- Implement SSH cert renewal in unixd
- Establish a good way for users to acquire the SSH CA's public key, possibly with a suggested `known_hosts` entry to copy/paste.

**Comment by Firstyear** [comment]
I will respond to this soon @Gabgobie but I've just been swamped lately. Thanks for taking the time to write this though :)

**Comment by markafarrell** [comment]
Just to add some extra information.

You can also distribute SSH CA's public key using the `authorized_keys` file (and as a result the `AuthorizedKeysCommand`) (https://manpages.debian.org/experimental/openssh-server/authorized_keys.5.en.html#cert-authority)

This makes is possible to limit the blast radius of compromised SSH CAs and to also easily rotate SSH CAs.

I have created a little playground that can be used to play around with ssh certificates to see how everything works.

https://github.com/markafarrell/ssh-cert-playground

I think it would be pretty easy to add support for distributing SSH CAs to kanidm (its essentially idential to how public keys are distrubuted currently) that would allow the use of an external SSH CA (e.g [Openbao](https://openbao.org/docs/secrets/ssh/signed-ssh-certificates/)) with `kanidm`.

In the future if `kanidm` itself implements an SSH CA it can re-use the same SSH CA distribution method.

**Comment by Gabgobie** [comment]
Thank you for getting involved and the amount of effort you put into your demonstration. To be able to get an answer out, I'll just trust that the CA per user via `AuthorizedKeysCommand` is possible without running your scripts. I did read them though!

To be able to do your comment justice, I think it would be valuable to get an understanding of your use-case. I'll try to outline some assumptions to give context for my thoughts.

Assumption 1: You want to use SSH certs.
Assumption 2: You want to integrate with Openbao
Assumption 3: Your server integrates with Kani using `unixd`

Status quo: This _should_ already be possible. When using an off-the-shelf solution, it likely expects to use one or very few `UserCAKey(s)` for your servers and users. You install the CA key into the `TrustedUserCAKeys` file and `unixd` handles the group assignments etc. To my understanding, there isn't really a need for action as you would probably instruct Kani to distribute the same key to every request regardless.

What changes with direct Kani support: Distribution would get centralized but you would need to touch the machines to install `unixd` anyway so the added value is questionable. It only really gets interesting when Kani _is_ the SSH CA but at that point why not drop the CA part, handle key validity internally and apply the restrictions to the individual keys?

We'd need to support adding the attributes listed in [the OpenSSH manpages you linked](https://manpages.debian.org/experimental/openssh-server/authorized_keys.5.en.html#AUTHORIZED_KEYS_FILE_FORMAT) to keys, at which point we would "accidentally" support SSH CAs but most importantly have the ability to communicate restrictions to `sshd`.

We'd be dodging the entire SSH CA mess and could give `unixd` a TTL in case of short-lived keys, validity windows if for example a key may only be used 9-5 etc. Outside this time frame, a key would not be returned when the `AuthorizedKeysCommand` is executed.

The only advantage of using a SSH CA I can see right now would be the offline capability. When all keys in `unixd` expire, there is no way to get in whereas a trusted CA may sign another short-lived key that will be accepted. The current solution is that keys are not restricted in their validity and revocation happens online. To achieve revocation against a signed key you don't know, you'd need to rotate the entire CA (hence the CA per user thought in the second to last section of my previous comment). Even if you know the key and revoke it, you'd need to deploy the KRL either manually or the servers would have to see Kani. Unless key revocation is online-only like it is now, there would be a good chance of missing a node when deploying the new KRL. Additionally, I personally find explicit distrust schemes not that pretty - It's a risky business to say "trust everything except these" when you can't prove that _everything_ is actually the set of items you think it is. Using plain SSH keys inverts this to say "Trust nothing except these" and _these_ is a set of keys you can audit as often as you'd like to prove that there are only the ones you expect. We could strongly suggest to add an SSH key without a bounded lifetime to each server as a break-glass access method. (If we can detect the use of this key, we should probably raise all sorts of alerts in the logs)

Edit: I think we can detect this at the very least via the [environment="NAME=value"](https://manpages.debian.org/experimental/openssh-server/authorized_keys.5.en.html#environment=_NAME=value_) attribute and check for the presence of `BREAK_GLASS_KEY_USED=true` or something more unique to avoid false positives.

---

If the goal is to hand short-lived keys to employees/colleagues/friends/other, you would still need to have some sort of client-side helper application (for `step` that would be `step-cli`, for us probably integrated in `kanidm-tools` unless that is considered too heavy for non-admin-workstations) that facilitates generating a key pair and getting it signed by the CA. Instead of signing a cert, it could just generate a key pair and hand it to Kani for distribution. If we want to be especially thorough and ensure that you aren't handing in a pubkey for someone else, we could send a challenge to be signed so you prove ownership of the private key.

The public key that was just handed in would then be be assigned restrictions similar to or more restrictive than the ones possible when using a certificate. These _could_ be echoed to the client as well to facilitate client-side renewal behavior, which is admittedly an advantage of certs - the client can inspect them and know about their validity.

---

What I'm ultimately wondering about is if there is anything SSH certs _can_ do that we _can't_ do (better) with `unixd` and plain SSH keys that makes them worth pursuing and taking on their flaws.

**Comment by Firstyear** [comment]
> What I'm ultimately wondering about is if there is anything SSH certs _can_ do that we _can't_ do (better) with `unixd` and plain SSH keys that makes them worth pursuing and taking on their flaws.

It's about being able to provide some form of exfiltration prevention.

SSH certs do this by being short lived - a stolen private key can only live for so long.

SSH Key Distribution via unixd relies on this through key storage in a secure enclave such as a yubikey. 

You can obviously mix these two for both properties. But the main thing is they protect from private key theft in different ways. Today with unixd alone, we can't "prove" if your key is hardware bound, but we do have the supporting code in place to be able to make that a policy if we wanted. We just haven't had time.

**Comment by Gabgobie** [comment]
> It's about being able to provide some form of exfiltration prevention.

It does quite the opposite though. It opens us up to having long/infinitely lived keys that we don't know of and can't block without revoking the entire CA.

With `unixd` on the other hand, we can assert strict and fine-grained validity rules for plain keys that affect which keys are returned when the `AuthorizedKeysCommand` is called.

---

Unless I got the wrong picture, Kani tries to protect everyone, including privileged users from shooting themselves in the foot.

Now let's imagine an admin at 4:30 AM trying to get into a server because something broke and the application rejects the valid cert (been there - incorrectly handled timezone). It's been a long day and said admin just wants to be done. The certificate issuer (Kani) will not issue a cert with validity constraints drastically deviating from the usual policy (backdating validity by multiple hours). Said admin may think it wise to just take the issuer key and manually sign a certificate - the result is a trusted key that cannot be revoked because it is not part of the audit trail. Worse yet, it may not even encode the not_after field and be valid forever.

Would a reasonable person do this? Probably not.
Have I seen worse? Yes!

This _cannot_ happen when we do our own thing instead of the horrible mess that is ssh certs.

So far I still only perceive the offline-capability of certs as something we can't easily imitate with `unixd`.

**Comment by yaleman** [comment]
> Said admin may think it wise to just take the issuer key and manually sign a certificate

If your environment lets an admin trivially grab the signing key and sign things then you're already doing it so wrong that I'm not sure we can do much about that. Yes, I've seen it, etc etc. 

The CA's also useful for where unixd use isn't possible, ie cert auth to Gitlab which has come up as a ssh-key-that's-not-for-normal-openssh-login use case in the past.

**Comment by Firstyear** [comment]
> > It's about being able to provide some form of exfiltration prevention.
> 
> It does quite the opposite though. It opens us up to having long/infinitely lived keys that we don't know of and can't block without revoking the entire CA.

Uhhh .... no? Most SSH ca's I'm aware of issue *short* lived certificates, not *long* ones. If you want to protect a longterm key, you need it to be hardware bound. 

> 
> This _cannot_ happen when we do our own thing instead of the horrible mess that is ssh certs.

Yes I would like to add ssh ca functionality, no today isn't that day. 

> 
> So far I still only perceive the offline-capability of certs as something we can't easily imitate with `unixd`.

ssh key distrubtion caches the keys offline, provided the user account was resolved (doesn't even need to have logged in to the machine). So provided you do "id user" on the machine *at least* once, then they can ssh in.

**Comment by Gabgobie** [comment]
> The CA's also useful for where unixd use isn't possible, ie cert auth to Gitlab which has come up as a ssh-key-that's-not-for-normal-openssh-login use case in the past.

I never considered this use-case. Sounds interesting. Not sure if SSH certs are the best solution here but also didn't look into it yet. It's not like `git+krb5` is supported just because GL does ^^

> If your environment lets an admin trivially grab the signing key and sign things then you're already doing it so wrong that I'm not sure we can do much about that. Yes, I've seen it, etc etc.

The example wasn't good but the point I was trying to make is that _someone_ will always have access to the CA key and that someone could trivially sign and exfiltrate a key whereas a plain ssh key would have to remain enumerable within the `AuthorizedKeys` file.

The only way to counter this would be frequent CA rotations which are probably doable within the realm of `unixd` integration but might be more difficult when integrating with other systems.

---

> > > It's about being able to provide some form of exfiltration prevention.
> > It does quite the opposite though. It opens us up to having long/infinitely lived keys that we don't know of and can't block without revoking the entire CA.
>
> Uhhh .... no? Most SSH ca's I'm aware of issue short lived certificates, not long ones. If you want to protect a longterm key, you need it to be hardware bound.

That's precisely the point I'm making. They are _supposed_ to be short lived but we can't _enforce_ that. There is no CA hierarchy. No CA lifetime - there isn't even a CA certificate, just a signing key.

The CA itself is a plain SSH key that is trusted to sign almost anything. You can pose some restrictions by prefixing it with the previously linked attributes but without a CA hierarchy you cannot really have a short-lived sub issuer to enforce short-lived leafs.

You also cannot enumerate issued leafs so you have to have absolute trust in the CA log.

> ssh key distrubtion caches the keys offline, provided the user account was resolved (doesn't even need to have logged in to the machine). So provided you do "id user" on the machine at least once, then they can ssh in.

Yup, I had some context in mind that I didn't write down. You are talking about the status quo whereas I was referring to the scenario in which we do our own thing rather than or in addition to SSH certs, which would give keys a TTL. As a result, if a host experienced issues causing it to lose line of sight to Kani, it could expire all keys whereas a CA would remain valid and still allow for short-lived keys to that host to be issued.

---

Either way. I don't mind SSH certs too much but think

- we have to be _very_ careful to handle them in a way that doesn't reduce assurances and
- that there is still _a lot_ of unused potential for plain ssh keys and `unixd`.

I think it would be a great first step to implement ssh key attributes. As demonstrated by @markafarrell, this would already provide support for distributing external SSH CAs but also enable us to restrict validity and force commands for any plain ssh key.

I can't commit to anything right now but could see myself looking into writing this part if there's no urgency (as seems to be the case) and of course you're interested.

Aside from that, we can extend the format with policies that `unixd` has to enforce. One example would be that a key is only valid from 9-5 on weekdays. This condition would be checked before a key is returned for the `AuthoriedKeysCommand`

The second we go beyond that, this will probably be getting a lot more complex.
